locator/network_topology_strategy.cc

```diff
@@ -336,12 +336,16 @@ future<size_t> network_topology_strategy::calculate_min_tablet_count(schema_ptr
     if (tablet_options.expected_data_size_in_gb) {
         tablet_count = std::max<size_t>(tablet_count, (tablet_options.expected_data_size_in_gb.value() << 30) / target_tablet_size);
     }
-    if (tablet_options.min_per_shard_tablet_count) {
-        tablet_count = std::max<size_t>(tablet_count, calculate_initial_tablets_from_topology(*s, tm, _dc_rep_factor, tablet_options.min_per_shard_tablet_count.value()));
-    }
-    if (tablet_count == 0) {
-        tablet_count = co_await calculate_initial_tablets_from_topology(*s, tm, _dc_rep_factor) * initial_scale.value_or(1);
+
+    auto min_per_shard_tablet_count = tablet_options.min_per_shard_tablet_count.value_or(
+            // If min_tablet_count is set, initial_scale should not be effective for
+            // compatibility with the deprecated "initial" tablet count.
+            (get_initial_tablets() || tablet_options.min_tablet_count) ? 0 : initial_scale.value_or(1));
+    if (min_per_shard_tablet_count) {
+        tablet_count = std::max<size_t>(tablet_count, co_await calculate_initial_tablets_from_topology(*s, tm, _dc_rep_factor,
+                                                              min_per_shard_tablet_count));
```

Comment:
After this change there should not be any call sites left with the default value for `min_per_shard_tablet_count`, so we can change the function signature to require it in all cases.

Reply:
And also drop `if (min_per_shard_tablet_count)` there,  We can even `SCYLLA_ASSERT(min_per_shard_tablet_count > 0.)` instead.

Reply:
> And also drop `if (min_per_shard_tablet_count)` there, We can even `SCYLLA_ASSERT(min_per_shard_tablet_count > 0.)` instead.

min_per_shard_tablet_count can still be 0 if `get_initial_tablets() || tablet_options.min_tablet_count`

Reply:
> > And also drop `if (min_per_shard_tablet_count)` there, We can even `SCYLLA_ASSERT(min_per_shard_tablet_count > 0.)` instead.
> 
> min_per_shard_tablet_count can still be 0 if `get_initial_tablets() || tablet_options.min_tablet_count`

But then we don't call calculate_initial_tablets_from_topology

Reply:
> > > And also drop `if (min_per_shard_tablet_count)` there, We can even `SCYLLA_ASSERT(min_per_shard_tablet_count > 0.)` instead.
> > 
> > 
> > min_per_shard_tablet_count can still be 0 if `get_initial_tablets() || tablet_options.min_tablet_count`
> 
> But then we don't call calculate_initial_tablets_from_topology

Right, I misunderstood. Well, that if is removed by later patches, so I don't think there's anything to do.

Reply:
okay.  Idealy, there was room for leaving FIXME comments documenting the intentions for that condition and also the std::max for the min per-shard tablet count that goes away in a later patch, to make it easier on the reviewers, but let's not hokd up this PR on that.